### DISCUSSION TITLE
Re-organise our building systems

### DISCUSSION THREAD

**Discussion by PsiACE** [databendlabs/databend#3419] [Ideas]
We now rely on the makefile as part of the build system to assist developers. 

- the model is relatively complex and we have to maintain cross-platform scripts, dependencies, and wrapped instructions. 
- this part lacks explanatory documentation telling developers how and why to do so, and some of the commands are even redundant.

## Goal

- reduce control over the developers' environment
- simplify workflow,only  keep the necessary and important commands
- add documentation to fill in possible details
- align the continuous integration with the build model for local environment

## What needs to be changed

- docker/
- scripts/
- tests/
- Makefile

**Comment by Xuanwo** [comment]
Is it a goal to remove all python scripts and rewrite them into a separate rust crate so that we don't need to depend on python?

**Reply by PsiACE** [reply]
This could be the result. :smile:

**Comment by ZhiHanZ** [comment]
related to issue https://github.com/datafuselabs/databend/issues/2647
we could make our test framework similar to other cloud native libraries such as prometheus.
pros of adopting this approach:
1. Uniform operation to run integration tests, and remove the need for a lot of CI scripts and cluster setup scripts
2. Remove python dependencies and mysql dependencies( run query through bendctl query "")
3. For environment setup, most of operation is finished through rust toolchain setup, which should be cleaner.

**Comment by Xuanwo** [comment]
Here is a bad case of our build system: https://github.com/datafuselabs/databend/pull/3499

Take check https://github.com/datafuselabs/databend/runs/4545281599?check_suite_focus=true for a look.

First of all:

```shell
nohup target/${BIN}/databend-meta --single=true --log-level=ERROR &
```

Hide the error of `databend-meta`, so our check can't exit at the correct time.

Then, the complex deploy/test logic makes it harder to figure what happened.

**Reply by PsiACE** [reply]
We really need a clean and unified programme.

**Reply by bohutang** [reply]
Please do it!

**Reply by ZhiHanZ** [reply]
I think it is part of task of test framework

**Reply by Xuanwo** [reply]
Is there a proposal for our test framework?

**Reply by ZhiHanZ** [reply]
https://github.com/datafuselabs/databend/issues/2647 , it is in design stage, and definitely will have a proposal before 0.7 release

**Comment by Xuanwo** [comment]
Also related to https://github.com/datafuselabs/databend/discussions/3518
